From: Jiri Slaby <jslaby@suse.cz>
Subject: x86-non-upstream-eager-fpu external modules fix
Patch-mainline: never, SUSE specific
References: bnc#1087086 CVE-2018-3665 bnc#1100091 bnc#1099598

This is a fix for:
patches.suse/x86-non-upstream-eager-fpu.patch

We silently broke kABI in that patch (kernel_fpu_begin/end are inlines)
and customers started seeing crashes due to added BUG_ON in
do_device_not_available. Fix this by soften it to a WARNING and do clts
to recover.

Note that the warning is printed in a ratelimit manner.

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 arch/x86/kernel/traps.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/arch/x86/kernel/traps.c
+++ b/arch/x86/kernel/traps.c
@@ -865,7 +865,10 @@ EXPORT_SYMBOL_GPL(math_state_restore);
 dotraplinkage void __kprobes
 do_device_not_available(struct pt_regs *regs, long error_code)
 {
-	BUG_ON(use_eager_fpu());
+	if (use_eager_fpu()) {
+		WARN(printk_ratelimit(), "Perhaps a KMP which was not rebuilt against the kernel protected against CVE-2018-3665 was loaded. Your system might be vulnerable, please rebuild all external modules. Trying to recover...");
+		clts();
+	}
 #ifdef CONFIG_MATH_EMULATION
 	if (read_cr0() & X86_CR0_EM) {
 		struct math_emu_info info = { };
